oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                            padding                            |
+|                         invocation_id                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         message_length                        |
```

Comment:
I'm confused -- is this really the higher-level message length, or just this frame length?

If the high-level message length, I'm not sure why it needs to be here.
If the frame length, it seems redundant with the body length; in fact I think it would be better to have the overall frame length instead of just the body length.

Reply:
I think the reason for putting the message length here was to preallocate a correct amount of bytes for the whole message once the first frame is received.

Reply:
This is really the message length. There are two reasons for storing the message length in the frame header. The first is as Ivan pointed out is to be able to reserve memory for the message so that we don't have to keep reallocating when we receive a new frame. The second is for frame validation.

Regarding the overall frame length vs the body length: we need both so we have to do a calculation anyway to determine the one that is not specified. The reason for choosing to specify the body length rather than the frame length is that it matches the approach in the virtio spec.